daemon/containerd/progress.go

```diff
@@ -0,0 +1,167 @@
+package containerd
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/containerd/containerd/content"
+	cerrdefs "github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/log"
+	"github.com/containerd/containerd/remotes"
+	"github.com/docker/docker/pkg/progress"
+	"github.com/docker/docker/pkg/stringid"
+	"github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/sirupsen/logrus"
+)
+
+type updateProgressFunc func(ctx context.Context, ongoing *jobs, output progress.Output, start time.Time) error
+
+func showProgress(ctx context.Context, ongoing *jobs, out progress.Output, updateFunc updateProgressFunc) func() {
+	stop := make(chan struct{})
+	ctx, cancelProgress := context.WithCancel(ctx)
+
+	var (
+		ticker = time.NewTicker(100 * time.Millisecond)
+		start  = time.Now()
+	)
+
+	for _, j := range ongoing.Jobs() {
+		id := stringid.TruncateID(j.Digest.Encoded())
+		progress.Update(out, id, "Preparing")
+	}
+
+	go func() {
+		defer func() {
+			ticker.Stop()
+			stop <- struct{}{}
+		}()
+
+		for {
+			select {
+			case <-ticker.C:
+				if !ongoing.IsResolved() {
+					continue
+				}
+				err := updateFunc(ctx, ongoing, out, start)
+				if err != nil {
+					logrus.WithError(err).Error("Updating progress failed")
+					return
+				}
+
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	return func() {
+		cancelProgress()
+		<-stop
+	}
+}
+
+func pullProgress(cs content.Store, showExists bool) updateProgressFunc {
+	return func(ctx context.Context, ongoing *jobs, out progress.Output, start time.Time) error {
+		pulling := map[string]content.Status{}
+		actives, err := cs.ListStatuses(ctx, "")
```

Comment:
The final length of `pulling` is known so the GC pressure can be reduced by providing a length hint when making the map.
```go
pulling := make(mapp[string]content.Status, len(actives))
```

Reply:
Done